### ISSUE TITLE
Is there any support for proxies?

### ISSUE THREAD

**Issue by qiujiangkun** [snapview/tungstenite-rs#177]
I can only access the dest host on another machine by ssh. If I can access it via proxy, it will be much convenient.
How can I do it for tungstenite?

**Comment by strohel** [comment]
Do you have an option to use [SSH TCP tunnelling](https://www.howtogeek.com/168145/how-to-use-ssh-tunneling/)?

**Comment by qiujiangkun** [comment]
Yes I can use it, but I have many different websocket addresses, so SSH TCP tunnelling is not very convenient.

**Comment by application-developer-DA** [comment]
We thought about making one of our internal libraries public (it brings support for proxies and supports such auth methods as basic, digest, ntlm, negotiate via SSPI / GSSAPI), but the support for proxies and their authentication methods should be part of another crate. We could introduce a `proxy` feature though that could rely on another crate for proxy and proxy authentication.

**Comment by PrivateRookie** [comment]
I created a branch that supports sock5 proxy in my forked repo, which may be helpful to you.
[pub fn connect_with_proxy](https://github.com/PrivateRookie/tungstenite-rs/blob/1d9289276518e5ab7e5194126d40b441d8938375/src/client.rs#L262)

**Comment by runapp** [comment]
Actually I guess support for http/https proxy will be more useful. As I write in https://github.com/snapview/tokio-tungstenite/issues/168, to inspect secure websocket connections, the best way is to use some mitm proxy software, which requires the client to send domain (HTTP GET request) rather than resolved ip address. And BTW it also requires support for specifying custom CA certificate, to allow mitm work.

For client behind firewall and only proxy allowed, there are many workarounds, such as using [gost](https://github.com/ginuerzh/gost) to convert almost any proxy protocol to a transparent one, as long as one don't need to inspect the raw stream or the target is plain, unsecured websocket.

**Comment by qiujiangkun** [comment]
Since we can use any stream that implementes `Read` and `Write`(or `Async` version), a proxy connection like [async_socks5](https://docs.rs/async-socks5/0.5.0/async_socks5/) would suffice.

**Comment by guotie** [comment]
good question, any solutions?

**Comment by siegfried** [comment]
It would be great to support system proxies like `reqwest` [does](https://docs.rs/reqwest/latest/reqwest/#proxies), which can be utilized by [Docker](https://docs.docker.com/network/proxy/) easily without change the code.

**Comment by application-developer-DA** [comment]
Yes, it could have been more convenient, but `reqwest` is by its nature a batteries included wrapper around HTTP, whereas `tungstenite-rs` is more like a core for support of WebSockets with optional blocking implementation for quick/typical use cases. `tungstenite-rs` is more akin to `hyper` in this regard rather than to `reqwest` if I can put it like this.

**Comment by siegfried** [comment]
> @siegfried what makes them so "system"? Is `HTTP_PROXY` a standard well-known env var that a lot of software expects to be set systemwide?

Supported by CURL, Docker, wget and reqwest. Call it "standard", "convention" or whatever you like. Not here for the definition debate. I agree with what @application-developer-DA said.

**Comment by Praying** [comment]
so, any solution can solve the proxy problems?

**Comment by qm3ster** [comment]
> so, any solution can solve the proxy problems?

@Praying depends on the type of proxy. Here's secure websocket over HTTP**S** proxy: https://github.com/websockets-rs/rust-websocket/issues/258#issuecomment-1052801655

**Comment by qm3ster** [comment]
In particular, since both that approach and [`async_socks5::connect`](https://docs.rs/async-socks5/latest/async_socks5/fn.connect.html) leave you with an un-wrapped `tokio::net::TcpSocket` or whatever else you had, you can easily handle `HTTP_PROXY` values of `http://` and `socks5://`, or lack thereof, without any `dyn`, enums, or generics.
Only what you do on the socket *before* passing it to  `tokio_tungstenite::client_async_tls` changes!

**Comment by lambdalisue** [comment]
FYI. I'm using [async-http-proxy](https://github.com/LinkTed/async-http-proxy) with [async-tungstenite](https://github.com/sdroege/async-tungstenite) and works well for general HTTP proxy servers such as nginx, squid, haproxy, etc.

**Comment by chuxi** [comment]
I resolved the problem with reference from `reqwest` proxy implementation and make up all the details in my post 
 [websocket](https://chuxi.github.io/posts/websocket/)

copy the `InnerProxy` and `ProxyStream` in your websocket client wrapper file. I did not supply it as library, because maybe the `tungstenite` will support the feature.

please use `tokio-tungstenite`

**Comment by zsl99a** [comment]
我很抱歉，但我从另一个视角打开了一个新的 issues。#406

**Comment by JakkuSakura** [comment]
There are many forks and articles. Why no one makes a PR?

**Comment by gakonst** [comment]
hi @daniel-abramov - i am also looking for the same feature upstream (vs having to work with [OAI's fork of the tungstenite libs](https://github.com/openai-oss-forks/tokio-tungstenite)) - can you share some guidance on what you'd want to get this work merged in? and i can do this properly w/o slop etc. (i appreciate if if there's sensitivity to ppl barging in the codebase and 1-shotting PRs so want to be respectful of it)

**Comment by agalakhov** [comment]
Technically speaking, proxy support belongs to a separate crate with only a thin layer of compatibility within Tungstenite. There are many kinds of proxies and authentication algorithms. They all are one OSI level deeper than Tungstenite.

**Comment by daniel-abramov** [comment]
Hey @gakonst. That's right, OpenAI maintains its own fork of `tungstenite` - last time I checked, there [were **two** major differences](https://github.com/openai-oss-forks/tungstenite-rs/tree/signal-plus-proxy?rev=4fffad30fe373adbdcffab9545e9e9bf4f2fc19f): proxy support (as you mentioned) **and** per-message deflate support based on the implementation that Signal used for their purposes. I'll address them separately:

1. Per-message deflate: this one has a long story (there were multiple implementations, one of which was even merged and then reverted back several years ago, then there were some PRs that had performance issues, up until Signal shared their implementation based on the other PR, though I'm not sure if it was benchmarked). **TL;DR:** [see my latest comment](https://github.com/snapview/tungstenite-rs/pull/426#issuecomment-3987850788).
2. Proxy support: there has never been a consensus regarding which proxies/proxy authentication methods we want to support. AFAIR, our stance was: let the user handle the proxy on their side, then pass the stream to `tungstenite` so that `tungstenite` handles the WebSocket part. Are there good reasons to embed proxy support inside `tungstenite`? There were some PRs in the past, though they had some issues and were abandoned. I have not checked [the latest PR yet](https://github.com/snapview/tungstenite-rs/pull/530) and whether it's the one used in the fork. 

Personal note: I have not yet verified whether the forked version of `tungstenite` used by OpenAI has any diffs from Signal's implementation. Nor have I checked what their proxy support is based on (whether it's a cherry-pick from one of the PRs or their own implementation on top of it). It's certainly good to know that the changes they incorporated seem to work well in their apps in the real world! To my knowledge, their fork is not used outside the Codex ecosystem though, allowing them to iterate very quickly. Since `tungstenite` is used by many crates and frameworks, it's important to ensure that large decisions/changes we make do not cause issues or unwanted regressions (performance or otherwise) for downstream crates.

Hope this clarifies things a bit.
